Return kernel argument objects instead of names from unique declarations (for #789)

TeranIvy opened the pull request:
This pull requests makes `unique_declarations` and `unique_declns_by_intent` in the `Invoke` class in `psyGen.py` return the lists and dictionaries of kernel argument objects instead of just names. Also, the code is updated to use `argument_type` instead of `datatype` name for API argument types to avoid confusing them with Fortran datatypes. The latter was also done for `unique_proxy_declarations` code in `DynInvoke` class.

The tests are updated accordingly, however I have not added compilation to the modified tests as that is done in PR #784 currently in review.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3271,17 +3275,23 @@ def _invoke_declarations(self, parent):
 
         for intent in FORTRAN_INTENT_NAMES:
             if self._int_scalars[intent]:
+                arg_list = [arg.declaration_name for arg in
```
Since it's only the names we use in these local lists (`self._real_scalars` and `self._int_scalars`) I suggest we ammend the code that sets them up so that it still only stores the names and not the Arg objects. We could also name those lists appropriately by adding the `_names` suffix, e.g. `self._real_scalar_names`.

TeranIvy replied:
Ok, I amended the `__init__` method to store only names and not the objects and I also renamed those dictionaries as suggested. Please note that this required a  `for` loop over Fotran intents for an invoke scalar arguments (similar to the one for the kernels).

TeranIvy replied:
Ok, I amended the `__init__` method to store only names and not the objects and I also renamed those dictionaries as suggested. Please note that this required a  `for` loop over Fotran intents for an invoke scalar arguments (similar to the one for the kernels).

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5014,37 +5028,43 @@ def __init__(self, alg_invocation, idx, invokes):
         '''
-        if datatype not in GH_VALID_ARG_TYPE_NAMES:
+        if argument_type not in GH_VALID_ARG_TYPE_NAMES:
             raise GenerationError(
```
Not your fault but since we're here: I think this should be an `InternalError` as it 'should never happen'.

TeranIvy replied:
Agreed. Changed.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -542,41 +543,49 @@ def schedule(self):
-        '''
-        if datatype not in VALID_ARG_TYPE_NAMES:
+
+        :param string argument_type: the type of the kernel argument for \
```
Not your fault but s/string/str/.

TeranIvy replied:
Yep, changed.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -542,41 +543,49 @@ def schedule(self):
+        :param string argument_type: the type of the kernel argument for \
+                                     the particular API for which the intent \
+                                     is required.
+        :param access: optional AccessType that the declaration should have.
```
Again, not your fault but please add a `:type access:` entry.

TeranIvy replied:
Heh, I knew I forgot something when updating docstrings. Added.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -542,41 +543,49 @@ def schedule(self):
+
+        '''
+        if argument_type not in VALID_ARG_TYPE_NAMES:
             raise GenerationError(
```
Please could you turn this into an Internal error (and update the doc string appropriately).

TeranIvy replied:
Turned and updated docstring.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -589,39 +598,43 @@ def first_access(self, arg_name):
-        :param string datatype: the type of the kernel argument for the \
-                                particular API for which the intent is \
-                                required
+        :param string argument_type: the type of the kernel argument for the \
```
s/string/str/

TeranIvy replied:
Corrected.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -589,39 +598,43 @@ def first_access(self, arg_name):
         '''
-        if datatype not in VALID_ARG_TYPE_NAMES:
+        if argument_type not in VALID_ARG_TYPE_NAMES:
             raise GenerationError(
```
Again, please could you make this an `InternalError`.

TeranIvy replied:
Made and updated docstring.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
-    assert args['inout'] == []
-    assert args['out'] == ['f1']
-    assert args['in'] == ['f2', 'm1', 'm2']
+    args_inout = [arg.declaration_name for arg in args['inout']]
```
I think you can leave this first one as it was? An empty list is an empty list :-)

TeranIvy replied:
Yep. This is now the second one (`args['out']`) due to changes in #784.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
+    assert args_inout == []
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == []
+    args_in = [arg.declaration_name for arg in args['in']]
```
These first two don't need to be changed as args['in'] and args['inout'] should both be empty.

TeranIvy replied:
Yep. Reverted.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
+    args_inout = [arg.declaration_name for arg in args['inout']]
+    assert args_inout == []
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == []
```
Again, no need to change the first two.

TeranIvy replied:
Yep. Reverted.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == ['op']
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == []
```
No need to change the first and last.

TeranIvy replied:
Yep. Reverted.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1679,68 +1687,98 @@ def test_dyninvoke_first_access():
+    args_out = [arg.declaration_name for arg in args['out']]
+    assert args_out == ['cma_op1']
+    args_in = [arg.declaration_name for arg in args['in']]
+    assert args_in == ['cma_op1', 'cma_opb']
```
Unless I'm missing something, we should not have `cma_op1` listed as having both intent 'out' and intent 'in'?

TeranIvy replied:
In the test example used here, `20.5_multi_cma_invoke.f90`, `cma_op1` is "read" arg in `columnwise_op_app_kernel_type` call and "write" arg in `columnwise_op_mul_kernel_type` call. This was about the only example I found with multiple CMA operator calls. I added comments in the test to explain this.

TeranIvy replied:
Yes, you're right and I think the problem is deeper than that. We have 
```
        # sum_args behave as if they are write_args from
        # the PSy-layer's perspective.
        write_args += sum_args
        # readwrite_args behave in the same way as inc_args
        # from the perspective of first access and intents
        inc_args += readwrite_args
```
so byt his logic checks for "write" args should reject anything else than `AccessType.WRITE, AccessType.SUM`, whereas checks for "inc" should reject anything else than `AccessType.INC, AccessType.READWRITE`. However, now we have
```
        for arg in inc_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access != AccessType.WRITE:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
and
```
        for arg in write_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access == AccessType.READ:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
No wonder Coverage shows red for these lines. I am happy to create an issue for this.

TeranIvy replied:
Yes, you're right and I think we need to revisit the logic there. We have
```
        # sum_args behave as if they are write_args from
        # the PSy-layer's perspective.
        write_args += sum_args
        # readwrite_args behave in the same way as inc_args
        # from the perspective of first access and intents
        inc_args += readwrite_args
```
There are first checks for "inc" and "readwrite" args 
```
        for arg in inc_args:
            ...
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access != AccessType.WRITE:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
            ...
```
and then for pure writers
```
        for arg in write_args:
            # For every argument that is written to by at least one kernel,
            # identify the type of the first access - if it is read
            # or inc'd before it is written then it must have intent(inout).
            # However, we deal with inc and readwrite args separately so we
            # do not consider those here.
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access == AccessType.READ:
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
```
However, I think "However, we deal with inc and readwrite args separately so we do not consider those here" is not fool-proof as there is a possibility that a "write" arg can have a  "read"-like first access access other than pure "READ"'. I tested this hypothesis with
```
        for arg in write_args:
            first_arg = self.first_access(arg.declaration_name)
            if first_arg.access in (AccessType.READ, AccessType.READWRITE, AccessType.INC):
                if arg not in declns["inout"]:
                    declns["inout"].append(arg)
```
Now 'cma_op1' is correctly placed in `inout` and `in` lists. 

I created an issue and referred it is as TODO in `psyGen` and this test.

arporter replied:
Yes please.

TeranIvy replied:
Created the issue and referred it accordingly.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3232,17 +3234,19 @@ class DynScalarArgs(DynCollection):
         super(DynScalarArgs, self).__init__(node)
 
         if self._invoke:
+            # Get a list of Invoke scalar arguments
```
Thanks for the suggestion, done.

TeranIvy commented:
First review comments addressed. Back to @arporter.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1803,13 +1823,13 @@ def test_dyninvoke_uniq_declns_intent_cma_ops():
     assert args_in == ['cma_op1', 'cma_opb']
 
-    assert LFRicBuild(tmpdir).code_compiles(psy)
-
```
Does this code not compile? If not, please could you reinstate this but commented out with an explanation.

TeranIvy replied:
This compilation line is for `test_dyninvoke_uniq_declns_intent_ops(tmpdir):` test. `test_dyninvoke_uniq_declns_intent_cma_ops():` did not have compilation as there is already a compilation check for "20.5_multi_cma_invoke.f90" in the `dynamo0p3_cma_test.py`. I added it here, though, to be on the safe side.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3240,25 +3240,27 @@ class DynScalarArgs(DynCollection):
-            # We have a Kernel stub, so create a list of scalar arguments.
-            self._real_scalars = {}
-            self._int_scalars = {}
             for intent in FORTRAN_INTENT_NAMES:
```
It was moved there. Are you talking about the `for` loop by Fortran intents where the values are initialised to empty lists? If so, I moved that as well, just below the dict initialisation. It seems to be needing both for the kernel stub.

arporter replied:
Oh yes. I was thinking that we could get rid of the `self._real/int_scalars = {}` but of course we can't. Apologies but would you therefore mind undoing this last change and putting the loop that initialises the various dict keys back so that it's only done for the kernel stub.

TeranIvy replied:
Nope, we can't get rid of that, I tried :) Undone the last change and put the key initialisation back.

TeranIvy commented:
Addressed the comments from 2nd review. I think the logic for declaring "write" args intents doesn't account for all possible "READ"-like accesses (see comment [above](https://github.com/stfc/PSyclone/pull/792#discussion_r442143446) and I am happy to create an issue for this.

Travis permitting, this can now go back to @arporter.

TeranIvy commented:
Addressed review comments and merged the latest master. Back to @arporter.

arporter commented:
All good now. (The two un-covered lines were not covered before and will be addressed as part of #801.)
